CIP-0144? | Wallet Connector API

nazrhom opened the pull request:
A wallet connector API. The goal of this CIP is to define a connection API that is independent of functionalities provided by wallet extensions. Future CIPs will introduce extensions to gain parity, and eventually surpass the capabilities offered by CIP-30 today.

See [CPS-10](https://github.com/cardano-foundation/CIPs/blob/master/CPS-0010/README.md) for motivation.

[(rendered version on fork)](https://github.com/nazrhom/CIPs/blob/nazrhom/full-data-wallet-connector/CIP-0144/README.md)

nazrhom committed:
Initial draft at wallet connector CPS

nazrhom committed:
Add query layer api

nazrhom committed:
Minor details

nazrhom committed:
Restructure CIP document

nazrhom committed:
Removing extra section

nazrhom committed:
Add versioning

nazrhom committed:
Minor refactors

nazrhom committed:
Add query layer API

nazrhom committed:
Fix according to review

nazrhom committed:
Remove CPS doc

Ryun1 commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,1601 @@
+---
+CIP: ?
+Title: Full-data wallet connector
+Status: Open
```
```suggestion
Category: Tools
Status: Proposed
```
Reorder header, fix status

Ryun1 commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,1601 @@
+CIP: ?
+Title: Full-data wallet connector
+Status: Open
+Category: Tools
```
```suggestion
```

Ryun1 commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,1601 @@
+  - Giovanni Garufi <giovanni@mlabs.city>
+Implementors: []
+Discussions:
+    - https://github.com/cardano-foundation/cips/pulls/?
```
```suggestion
    - https://github.com/cardano-foundation/cips/pulls/957
```

nazrhom committed:
Update CIP-XXXX/README.md

Co-authored-by: Ryan <ryan.williams@intersectmbo.org>

nazrhom committed:
Update CIP-XXXX/README.md

Co-authored-by: Ryan <ryan.williams@intersectmbo.org>

nazrhom committed:
Update CIP-XXXX/README.md

Co-authored-by: Ryan <ryan.williams@intersectmbo.org>

nazrhom committed:
Update CIP-XXXX/README.md

Co-authored-by: Ryan <ryan.williams@intersectmbo.org>

rphair commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,1601 @@
+    - https://github.com/cardano-foundation/cips/pulls/957
+Created: 2024-12-13
+License: CC-BY-4.0
+Version-Connection-API: 0.0.0
+Version-CIP-30-Extension: 0.0.0
```
```suggestion
```
@nazrhom - at this time, we have no such headers in the CIP front matter.  Adding them could be ignored by some parsers (not all of which are YAML parsers: some have just autistically cut up the text) and generate errors on others.

Until we get feedback from the owners of https://github.com/cardano-foundation/cf-cip-frontend we will have to keep these fields to the ones defined here: https://github.com/cardano-foundation/CIPs/blob/master/CIP-0001/README.md#header-preamble

rphair replied:
p.s. I did also meant to say that, in my opinion, what you've written in the (appreciated) Versioning section describes the version state & progression adequately.

nazrhom replied:
Thanks @rphair I have created a separate table in the versioning section to keep track of those

rphair replied:
yes, that's exactly what we are looking for 👍 though I might suggest adding a header line to that table (or removing the line with the empty column headers)

rphair commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,1601 @@
+License: CC-BY-4.0
+Version-Connection-API: 0.0.0
+Version-CIP-30-Extension: 0.0.0
+---
```
```suggestion
Solution-To: CPS-0010
Solution-To: CPS-0012
---
```
On that note, now that we are gradually evolving better mapping between CIP and CPS sets, we should make it official in the header.

@Ryun1 @perturbing @Crypto2099 I think we also need to establish a policy on how these inclusions should be performed in the other direction... I would propose this CIP also modify the target CPS to add the corresponding header line there: which would risk more merge conflicts but provide a more robust CIP repository in the long run (since these mappings would be less likely to get missed if adding one also added the other).

@Crypto2099 mentioned at yesterday's meeting that this could be a solution to more than one CPS: if there are others please by all means add them (also we could remove CPS-0012 if we're not all in agreement that this CIP "solves" that CPS).

nazrhom replied:
I will add the `Solution-to: CPS-0010` header, as for CPS-12, maybe we should add the `Solution-to` tag to https://github.com/cardano-foundation/CIPs/pull/869 ?

rphair replied:
yes, thanks: suggested in https://github.com/cardano-foundation/CIPs/pull/869#discussion_r1909022087.

rphair replied:
@nazrhom regarding the applicability of this CIP as a solution to CPS-0012: doesn't it provide an answer to the "open question" https://github.com/cardano-foundation/CIPs/tree/master/CPS-0012#how-can-we-encourage-wallet-developers-to-adopt-the-solution ?  I would like to hear what you & @Ryun1 @Crypto2099 @perturbing think about how broadly we should be thinking in matching these CIPs to CPSs.

rphair commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,1601 @@
+  - Adding a query layer extension API to enable a full data wallet
+
+
+## Path to Active
+
+- Implementing a wrapper that takes a CIP-30 compatible API and transforms it to be compatible with CIP-XXXX.
+- This CIP depends on CIP-139 to be active, in particular wallets need to support the queries required by the `cip-139` extension.
+- Two or more wallets implement support for this CIP.
```
```suggestion
## Path to Active

- Implementing a wrapper that takes a CIP-30 compatible API and transforms it to be compatible with CIP-XXXX.
- This CIP depends on CIP-139 to be active, in particular wallets need to support the queries required by the `cip-139` extension.
- Two or more wallets implement support for this CIP.
```
@nazrhom given the care you have taken with the rest of the spec I don't want to rewrite this for you: but we will need to get 2 separate lists of items to achieve `Active` status vs. items to validate the CIP itself:
### Acceptance Criteria
### Implementation Plan
The distinction is explained here & please let's work together as necessary to get 2 separate lists as per https://github.com/cardano-foundation/CIPs/blob/master/CIP-0001/README.md#path-to-active 🙏

nazrhom replied:
Thanks for pointing this out @rphair I must have deleted those sections from the template by mistake and didn't realise.

I have now split the points into the two sections.

rphair replied:
Fixed by https://github.com/cardano-foundation/CIPs/pull/957/commits/8040b5c3813e7c134172a77b2aaf339b55177786.

rphair reviewed (commented):
At initial review in the CIP meeting yesterday we agreed this represents long-awaited progress as a wallet API that we have needed for some time.  We feel optimistic this will move ahead after some refinement so agreed to accept its CIP candidate status immediately.

@Ryun1 @Crypto2099 @perturbing we should start tagging wallet devs left & right on this, as well as kicking off some activity in the CIP Discord in both the Wallets and Query Layer threads.  - Q: is the Wallets Working Group still meeting, and could they start chewing on this there?

In the meantime @nazrhom please rename the containing directory to `CIP-0144` and update the link to your proposal in the original posting with the new pathname. 🎉

nazrhom committed:
Add CIP number and remove extra version headers from top-level table

nazrhom committed:
Merge branch 'nazrhom/full-data-wallet-connector' of github.com:nazrhom/CIPs into nazrhom/full-data-wallet-connector

nazrhom committed:
Fix verion table formatting

nazrhom committed:
Add Solution-To tag

rphair commented on CIP-0144/README.md:
```diff
@@ -0,0 +1,1604 @@
+
+While the CIP is in preparation, these versions shall be set to `0.0.0`. The moment this CIP is merged the versions shall be set to `1.0.0`, and all implementations should consider that the current version. Any changes to either API should come in form of PRs to this CIP.
+
+|     |    |
+| --- | --- |
+| Version-Connection-API | 0.0.0 |
+| Version-CIP-30-Extension | 0.0.0 |
```
```suggestion
| API | Version |
| --- | --- |
| Version-Connection-API | 0.0.0 |
| Version-CIP-30-Extension | 0.0.0 |
```
... follow up to https://github.com/cardano-foundation/CIPs/pull/957#discussion_r1908650806 & https://github.com/cardano-foundation/CIPs/pull/957/commits/1b4e5b36837d9f67a3cea8052962ea52319eaf69, to either give column headers to this table... or the header bar could be removed by cutting the first two lines.

nazrhom committed:
Update CIP-0144/README.md

Co-authored-by: Robert Phair <rphair@cosd.com>

nazrhom committed:
Split Path to active into two sections

nazrhom committed:
Merge branch 'nazrhom/full-data-wallet-connector' of github.com:nazrhom/CIPs into nazrhom/full-data-wallet-connector

Ryun1 commented on CIP-0144/README.md:
````diff
@@ -0,0 +1,1610 @@
+
+We will use the following schema to define operations:
+
+```
````
```suggestion
```json
```

Ryun1 commented on CIP-0144/README.md:
```diff
@@ -0,0 +1,1610 @@
+
+We will reference several JSON schemas throughout the document, these are:
+
+- [cip-116](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0116) which provides a JSON encoding of Cardano ledger types. Note that this CIP defines a schema for each ledger era. When referring to a type from this schema we refer to an `anyOf` of all the schemas in which that type is defined. *[Maybe we should only reference the latest era and update the CIP in the future? our current  ...
```
```suggestion
- [CIP-116 | Standard JSON encoding for Domain Types](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0116) which provides a JSON encoding of Cardano ledger types. Note that this CIP defines a schema for each ledger era. When referring to a type from this schema we refer to an `anyOf` of all the schemas in which that type is defined. *[Maybe we should only reference the latest era and update the CIP in the future? our current definition requires to be perpetually backward compatible with old ledger types]*
```
maybe a little nicer for navigation

Ryun1 commented on CIP-0144/README.md:
```diff
@@ -0,0 +1,1610 @@
+
+### Versioning
+
+In this CIP we are defining two different APIs: the connection API for wallets, and the CIP-30 *[maybe this should have another name to prevent confusion?]* extension which enables an own-data wallet. These two are separate components, below there is a table with separate entries for the versions of the connection API and CIP-30 Extension respectively. 
```
I know we already have a lot of CIPs/CPSs for this, and i don't want to increase work on this unnecessarily

It may make sense to have this proposal introduce the connection API, as I believe this API is agnostic of type of wallet(?)
AND then a separate proposal which adds the support/extension for the CIP30 functionality 

does this sound reasonable?

nazrhom replied:
Hey @Ryun1 I think that is a fair point. One argument that can be made against it is that we could split this even further. It is something that I was thinking quite a bit about when writing this document, in a way you could argue that the 5 points listed [here](https://github.com/cardano-foundation/CIPs/pull/957/files#diff-77fa7d71c1149f727dc693fb12eb68c95af075b7cc947bcb6bf9a673682d6ff7R72-R76) should all be standalone CIPs.
Having said this, I think it does make sense to have it separate, if you and other reviewers agree on it I am happy to separate them. My initial thought would be to have a CIP which introduces the transport agnostic specification, and use that to define the connection API. Then we could have one (or two?) other CIPs to define the CIP-30 and CIP-139 extensions. What are your thoughts? cc @rphair

nazrhom committed:
Update CIP-0144/README.md

Co-authored-by: Ryan <ryan.williams@intersectmbo.org>

nazrhom committed:
Update CIP-0144/README.md

Co-authored-by: Ryan <ryan.williams@intersectmbo.org>

nazrhom committed:
Remove extensions and leave wallet connector API

nazrhom committed:
Keep only wallet connection API

nazrhom commented on CIP-0144/README.md:
```diff
@@ -0,0 +1,354 @@
+
+### Versioning
+
+In this CIP we define a semantic versioning (SemVer) scheme that will be used by the connection API. Extensions defined in other CIPs (e.g. [CIP-XXX | Own data wallet](link)) must follow the same versioning schema.
```
```suggestion
In this CIP we define a semantic versioning (SemVer) scheme that will be used by the connection API. Extensions defined in other CIPs (e.g. [CIP-XXX | Own data wallet](https://github.com/cardano-foundation/CIPs/pull/986)) must follow the same versioning schema.
```

nazrhom commented on CIP-0144/README.md:
```diff
@@ -0,0 +1,354 @@
+
+There are two CIPs that are connected with this one:
+
+- [CIP-XXX | Own data wallet](link) which defines an extension for an [own-data wallet](https://github.com/cardano-foundation/CIPs/blob/master/CPS-0010/README.md#own-data-wallets) which is equivalent to what CIP-30 offers today
```
```suggestion
- [CIP-XXX | Own data wallet](https://github.com/cardano-foundation/CIPs/pull/986) which defines an extension for an [own-data wallet](https://github.com/cardano-foundation/CIPs/blob/master/CPS-0010/README.md#own-data-wallets) which is equivalent to what CIP-30 offers today
```

nazrhom committed:
Update CIP-0144/README.md

nazrhom committed:
Update CIP-0144/README.md

nazrhom committed:
Merge branch 'nazrhom/full-data-wallet-connector' of github.com:nazrhom/CIPs into nazrhom/full-data-wallet-connector

rphair commented:
@nazrhom when your commits have stabilised please post to let us know.  If you're ready, we might get it on the agenda for this meeting (https://hackmd.io/@cip-editors/106) where https://github.com/cardano-foundation/CIPs/pull/986 would also be introduced (hopefully by _you_).  We might choose a wallet emphasis for that meeting and try to invite a number from the working group to ensure we can start filling out review on GitHub for these related proposals.

nazrhom commented:
Hey @rphair the commits should have mostly stabilised by now. I split the CIP into 2 different ones as suggested (and integrated the CIP-139 extension directly in CIP-139). Could you please remind me when the next meeting is again? The link you send does not seem to work for me :(

Ryun1 commented on CIP-0144/README.md:
````diff
@@ -0,0 +1,356 @@
+
+##### Enable
+
+```
````
```suggestion
```json
```

nazrhom replied:
I'll add this across all CIPs

Ryun1 replied:
awesome thanks!

Ryun1 commented on CIP-0144/README.md:
````diff
@@ -0,0 +1,356 @@
+}
+```
+
+This is the entrypoint to start communication with the user's wallet. The wallet should request the user's permission to connect the web page to the user's wallet, and if permission has been granted, the APIs will be made available for the dApp to use. The wallet can choose to maintain a whitelist to not necessarily ask the user's permission every time access is requested, but this behavior is up ...
````
> but this behavior is up to the wallet and should be transparent to web pages using this API.

I like this a lot, this gives wallets the freedom to decide

Ryun1 commented on CIP-0144/README.md:
```diff
@@ -0,0 +1,356 @@
+
+When requesting the functionalities of a cip extension, dApps can optionally specify a version number for it. Every cip extension defined in the future, must also define a versioning scheme following SemVer. When the version argument is not specified, wallets should take that as the greater version they implement. In this context - and for the rest of this discussion - we will assume versions are ...
+
+- the same major value as the one requested by the dApp AND
```
do we need to include SemVer versioning for extensions?
I feel like this maybe adding unneeded complexity

Existing CIP30 extensions, aren't versioned, instead to replace one you must provide another CIP
I think this is mostly fine

rphair replied:
In prior discussions for both CIPs and CIP-0030 extensions for last 4 years or so, in my recollection all parties agreed each time that an integer "serial number" was both adequate and future-proof as a version ID _except_ for those who argued that "semantic versioning is _always_ preferred" as a general principle.

mkazlauskas replied:
I think we should have the option to update an API through versions. It enables a more 'agile' approach to defining and implementing CIPs

Ryun1 commented on CIP-0144/README.md:
```diff
@@ -0,0 +1,356 @@
+Returns true if the dApp is already connected to the user's wallet, or if requesting access would return true without user confirmation (e.g. the dApp is whitelisted), and false otherwise.
+
+
+##### SupportedExtensions
```
does connection have to be enabled to be able to call this method?

Ryun1 commented on CIP-0144/README.md:
```diff
@@ -0,0 +1,356 @@
+
+This appendix contains additional schemas for types that are used in the APIs.
+
+#### Connection API Data Types
```
really nice idea to include these :)

Ryun1 commented on CIP-0144/README.md:
````diff
@@ -0,0 +1,356 @@
+We will use the following schema to define operations:
+
+```json
+{
````
```suggestion
{
  "connectionVersion": { "type": "semVar" },
```

does it make sense to have the connection API versioned?

nazrhom replied:
I think so, I added an endpoint both to the connection API and to the different extensions to report their `apiVersion`. Adding it here however, would potentially make each endpoint have it's own separate version which might be a bit too much complexity

Ryun1 commented on CIP-0144/README.md:
````diff
@@ -0,0 +1,356 @@
+```
+{
+  "operation": "enable",
+  "request": { "$ref": "#/appendix/Extensions" },
````
just a thought, would there be a usecase where a dApp doesnt request a whole extension?
perhaps a dApp only wants a couple methods, maybe this would just add unneeded complexity

Ryun1 replied:
Discussed in meeting
we can consider this in the future, or when we have more feedback from stakeholders

Ryun1 commented on CIP-0144/README.md:
```diff
@@ -0,0 +1,356 @@
+Returns the API version for the wallet connection API. This must correspond to the value of `Connection-API` specified in this document, appropriately transformed into a `SemVer` object.
+
+### Versioning
+
```
note to discuss using SemVar or other versioning scheme

rphair commented:
@nazrhom now that the last hour's CIP meeting has confirmed we have a new title taxonomy `Wallet Connector -` for CIP-0144 compliant extensions, as per our meeting discussion could you also add a table — initially populated with
- https://github.com/cardano-foundation/CIPs/pull/986

as @Ryun1 has done here for CIP-0030? 🙏
- https://github.com/cardano-foundation/CIPs/pull/980

@Ryun1 @Crypto2099 and other wallet reviews... I would leave it to your consensus about whether or not "candidate" (not yet merged) CIPs should be included here.
- We still have some candidate CIP-0030 extensions (not "accepted" as per the title of the table) that aren't included on the table (https://github.com/Ryun1/CIPs/blob/cip30-extension-discoverability/CIP-0030/extensions-register.md) so — if we are purists and don't include PRs / candidates — then perhaps the table should be blank for now. 🤔

Ryun1 commented on CIP-0144/README.md:
```diff
@@ -0,0 +1,356 @@
+---
+CIP: 144
+Title: Wallet Connector API
```
```suggestion
Title: Extensible Wallet Connector Framework
```
I feel maybe we can make the title a bit more specific
This CIP is defining a new framework for wallet connectors, whilst also proposing a connector API

rphair replied:
yes, that's good :rocket: but I still think the shorter form (`Wallet Connector -`) should be used for its compliant extensions (e.g. https://github.com/cardano-foundation/CIPs/pull/986#pullrequestreview-2624359783).

mpizenberg commented:
Just to say 👋 

This is a subject I’m super interested in, and really want to help improve both sides of the equation: user UX and dapps capabilities. I would be happy to contribute and help restart the discussions on this subject in a couple months. I’m currently busy on a couple things, but this is super important and will be critical for adoption and retention, both of users and devs.

Ryun1 commented:
@mpizenberg

im also keen to move this along
happy to collaborate on this when possible 💪

nazrhom commented:
Hi @Ryun1 @rphair I am sorry for not replying here sooner but have been busy with other work. Good news is I have received approval from MLabs to spend some time pushing this CIP forward: addressing comments, making the required changes, etc. So I am keen to pick this up. I am wrapping up another project but should be able to start making some progress here towards the end of August.

@mpizenberg I would be happy to collaborate with you on this!

nazrhom commented:
@mkazlauskas Thank you very much for your feedback on the PR, I am sorry for the incredibly late reply. I will reply to each point in individual comments, but just wanted to thank you for looking into this!

rphair commented on CIP-0144/README.md:
```diff
@@ -0,0 +1,356 @@
+
+While this CIP attempts to define an API in a transport agnostic way, implementations will be forced to pick a specific transport. In the spec we have not given specifics on how to namespace and generally structure the api in an implementation. This is both because details depend on the underlying transport that is chosen to implement the API, but also because we wanted to make backwards compatib ...
+
+Each transport specific implementation will make a choice on how to namespace access to the operations. In the following we give a description of how this should work for an implementation using an injected javascript object. Support for different transports can be added to this CIP in form of PRs.
```
@mkazlauskas @nazrhom this is a question I would hope to have @Ryun1's insight about.

Ryun1 replied:
> The argument against this would be to avoid opening too many interconnected CIPs

i feel that is probably okay
since it would be quite nice if this CIP was transport-agnostic

nazrhom committed:
Update CIP-0144/README.md

Co-authored-by: Ryan <ryan.williams@intersectmbo.org>

rphair commented on CIP-0144/README.md:
```diff
@@ -0,0 +1,356 @@
+    - https://github.com/cardano-foundation/cips/pulls/957
+Created: 2024-12-13
+License: CC-BY-4.0
+Solution-To: CPS-0010
```
yes please... but without the hyphen 😅 as in this pending example:
* https://github.com/cardano-foundation/CIPs/pull/986

Ryun1 commented on CIP-0144/README.md:
````diff
@@ -0,0 +1,356 @@
+```
+{
+  "operation": "enable",
+  "request": { "$ref": "#/appendix/Extensions" },
````
i feel that one Cardano network at a time, is probably okay
and matches existing implementations

mkazlauskas replied:
Agree, one network should be enough for now. Although it would be interesting to learn about Hydra use cases

Ryun1 commented:
@mpizenberg 

> where can I find links to join the call (if that’s possible at all)?

of course!
All meetings are via the CIPs discord ([see invite](https://discord.com/invite/7JCs24RY))

meetings are every other Tuesday at 4pm UTC
[see invite for tomorrow's call](https://discord.com/invite/F7J8zFcX?event=1370460084454293625)

mkazlauskas commented on CIP-0144/README.md:
```diff
@@ -0,0 +1,356 @@
+
+### CPS-10
+
+[CPS-0010](https://github.com/Ryun1/CIPs/blob/cps-wallet-connector/CPS-0010/README.md) discusses several limitations of CIP-30. In this CIP we try to solve the issues pointed out in that CPS. There are still some areas to improve on: things like the event listener API are intentionally left out of this CIP to prevent bloat of scope. We welcome future CIPs, or updates to this CIP to refine any lim ...
```
I was thinking that extensions:
- would be defining their schemas for base operation payloads (`request`/`response`/`error`)
- probably shouldn't be adding new base operations, as it adds new requirements for transport CIPs/implementations

Given this CIP (and naming of `request`/`response` operations), transport specifications would probably imply only a single value resolved from a 'request' - a typical pull-based API structure, e.g. injected javascript object transport implementation would probably map requests/responses into a promise based API like `api.cip123.operation(r: Request) => Promise<Response>`

If you wanted to add an extension for events, you couldn't really do it by adding an `on` 'request', because transports would generically expect it to resolve with only a single value like all other requests.

Am I missing something - could you create an extension for events given this base CIP?

nazrhom committed:
Update CIP-0144/README.md

Add extensions table

nazrhom committed:
Update CIP-0144/README.md

Ryun1 commented on CIP-0144/README.md:
```diff
@@ -0,0 +1,362 @@
+- `Refused`: (-3) The request was refused due to lack of access - e.g. wallet disconnects.
+- `AccountChange`: (-4) The account has changed. The dApp should call wallet.enable() to reestablish connection to the new account. The wallet should not ask for confirmation as the user was the one who initiated the account change in the first place.
+- `NotSatisfiable`: (-5) The request is structurally correct, but the wallet can not satisfy it for some reason.
+
```
would a new error be needed 
for the case where a dApp requests a network at enable time
but the wallet is not connected to it?

nazrhom committed:
Update CIP-0144/README.md

Co-authored-by: Ryan <ryan.williams@intersectmbo.org>

nazrhom committed:
Update CIP-0144/README.md

Co-authored-by: Robert Phair <rphair@cosd.com>

nazrhom committed:
Update README.md

Add chainId to enable call. Also add new error type if the requested network is not avialable

rphair commented:
@nazrhom I'm trying to find how we might progress this & its related CIP candidate, now flagged `Waiting for Author` for some of the same reasons as this one:
* https://github.com/cardano-foundation/CIPs/pull/986#issuecomment-5311131323

I remember around this time last year the candidate CIP-0144 was being hailed "the next great leap forward" at CIP meetings and the like.  Then 11 months ago the development & dialogue here simply stopped.
* @Ryun1 & other wallet advocates: was there any difference of opinion at the time that caused this to be stalled?  Did some other development supersede it somehow?
* @nazrhom we realise changes in professional or personal circumstances might simply have halted your efforts: if this is the case, please kindly post here to confirm.

I recall the purported benefits of having such a wallet API but I would have to backtrack this understanding & expectation if indeed we were unable to produce a CIP for it:
* Conversely: we should do whatever it takes to get this back on track towards merge if & when the wallet community demonstrates support & this candidate can be declared complete.
  * @Ryun1 you had a tremendous amount of feedback incorporated into this proposal: likewise I think the editors would accept your own judgement about whether this candidate would produce usable implementations & could perhaps be merged "as-is" or as a disclaimed "best effort".
* Alternatively: if there's no further comment from the `Waiting for Author` state then this would be progressively moved toward closure: since proposals like this are of no value _unless we can document a favourable peer review_.

Coincidentally we are just about to dive into long standing CIP repository issues, in which wallet issues are predominant.  If this PR is still a solution to long pending UX or DevX problems then we should be are of this ASAP in that effort: so please anyone with information on this subject should feel free to post.

cc (reviewers so far) @mkazlauskas @mpizenberg

nazrhom commented:
Hi @rphair 

Unfortunately you are correct. I am not currently working in the Cardano ecosystem anymore, and as such don't have enough time to bring this to conclusion. 
I want to thank you, @Ryun1 and all the other reviewers for all the help and guidance. I will definitely keep an eye on how this progresses, and am happy to offer feedback on any further changes that should be required.

rphair commented:
Thanks @nazrhom for the clarification.  @Ryun1 I would recommend that we keep this open for perhaps a month — considering that there are still summer holiday absences — to either:
* merge if you deem it fit (maybe after a few commits to make it so); or (after that waiting period, if not)
* close it in favour of some future evolution of the wallet interface that reflects needs & perceptions at that time.

@Crypto2099 may be coming back on board for some measure of CIP review, so this would also give him some time to either conclude this, if interested, or recommend closure.

In any case I think the most important question is _Could this CIP candidate get the endorsement of wallet representatives at this point, and would participation in the proposed standard continue?_  Without that confirmation we would unfortunately have to consider this proposition & its related extensions abandoned.
